### PR TITLE
udp: add support for UDP connected sockets

### PR DESCRIPTION
**Pull request by santigimeno**
Following the conversation at https://github.com/libuv/leps/pull/10, I've implemented the API as @saghul suggested. Also I've targeted the PR to `v1.x` but could be easily retargeted to `master`.

I've split it into 3 different commits:

- unix: refactor getsockname/getpeername methods
- win,udp: allow to use uv_udp_open on bound sockets
- udp: add support for UDP connected sockets
    
    Add two new methods:
    
    `uv_udp_connect()` to connect / disconnect an UDP handle.
    `uv_udp_getpeername()` to get the remote peer address of a connected UDP
    handle.
    
    Modify `uv_udp_send()` and `uv_udp_try_send()` to accept a `NULL` `addr`
    to send messages over an "UDP connection".

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/913/

### CODE REVIEW

File: docs/src/udp.rst [hunk 1]
```diff
@@ -150,6 +150,45 @@ API
...
+    On unconnected handles, it returns `UV_NOTCONN`.
+
+    :param handle: UDP handle. Should have been initialized with
+        :c:func:`uv_udp_init` and bound.
+
+    :param name: Pointer to the structure to be filled with the address data.
+        In order to support IPv4 and IPv6 `struct sockaddr_storage` should be
+        used.
+
+    :param namelen: On input it indicates the data of the `name` field. O
```
Comment (cjihrig) [inline docs/src/udp.rst]: O -> On?

File: docs/src/udp.rst [hunk 2]
```diff
@@ -247,6 +286,12 @@ API
     (``0.0.0.0`` or ``::``) it will be changed to point to ``localhost``.
     This is done to match the behavior of Linux systems.
 
+    For connected UDP handles, `addr` must be set to `NULL`, otherwise it will
+    return `UV_EISCONN` error.
+
+    For connection-less UDP handles, `addr` cannot be `NULL`, otherwise it will
```
Comment (cjihrig) [inline docs/src/udp.rst]: No hyphen in connection-less (here and throughout).

File: src/unix/udp.c [hunk 3]
```diff
@@ -383,6 +388,51 @@ static int uv__udp_maybe_deferred_bind(uv_udp_t* handle,
...
+    return UV__ERR(errno);
+
+  handle->flags |= UV__UDP_CONNECTED;
+
+  return 0;
+}
+
+
+int uv__udp_disconnect(uv_udp_t* handle) {
+
```
Comment (cjihrig) [inline src/unix/udp.c]: No blank line.

File: src/unix/udp.c [hunk 4]
```diff
@@ -459,9 +514,13 @@ int uv__udp_try_send(uv_udp_t* handle,
...
 
-  err = uv__udp_maybe_deferred_bind(handle, addr->sa_family, 0);
-  if (err)
-    return err;
+  if (addr) {
+    err = uv__udp_maybe_deferred_bind(handle, addr->sa_family, 0);
+    if (err)
+      return err;
+  } else {
+    assert(handle->flags & UV__UDP_CONNECTED);
```
Comment (cjihrig) [inline src/unix/udp.c]: Does this need to be an assertion, or could we just return an error?

File: src/uv-common.c [hunk 5]
```diff
@@ -284,24 +284,87 @@ int uv_tcp_connect(uv_connect_t* req,
...
+  if (addr == NULL && !(handle->flags & UV__UDP_CONNECTED))
+    return UV_EDESTADDRREQ;
+
+  if (addr != NULL) {
+    if (addr->sa_family == AF_INET)
+      addrlen = sizeof(struct sockaddr_in);
+    else if (addr->sa_family == AF_INET6)
+      addrlen = sizeof(struct sockaddr_in6);
+    else
+      return UV_EINVAL;
```
Comment (cjihrig) [inline src/uv-common.c]: Don't we have some platforms where the errors are positive? I could be remembering incorrectly, but I thought that came up during the [`UV__ERR()`](https://github.com/libuv/libuv/commit/89cbbc895bb459c6ab5319fa86a32a9fecbf8744) discussion.

File: src/win/udp.c [hunk 6]
```diff
@@ -895,6 +909,51 @@ int uv__udp_bind(uv_udp_t* handle,
...
+    return uv_translate_sys_error(err);
+
+  handle->flags |= UV__UDP_CONNECTED;
+
+  return 0;
+}
+
+
+int uv__udp_disconnect(uv_udp_t* handle) {
+
```
Comment (cjihrig) [inline src/win/udp.c]: No blank line.

File: src/unix/udp.c (hunk 4 above)
Comment (santigimeno) [reply src/unix/udp.c]: I added the assertion as this should never happen. See: https://github.com/santigimeno/libuv/blob/94adb2020490b9774f7b1c12bf4bf7745dd2be72/src/uv-common.c#L338-L339

File: src/uv-common.c (hunk 5 above)
Comment (santigimeno) [reply src/uv-common.c]: `UV_EINVAL` is negative on every platform, or you refer to something else?

Comment (cjihrig) [reply src/uv-common.c]: I was under the impression that that was not the case based on https://github.com/libuv/help/issues/39 and https://github.com/libuv/libuv/pull/1687. Isn't that the reason for this code - https://github.com/libuv/libuv/blob/0cdb4a5b4b706d0e09413d9270da28f9a88dc083/include/uv-errno.h#L26-L30?

Comment (santigimeno) [reply src/uv-common.c]: Considering that this is the definition of `UV_EINVAL`
https://github.com/libuv/libuv/blob/0cdb4a5b4b706d0e09413d9270da28f9a88dc083/include/uv-errno.h#L168, isn't it implied that `UV_EINVAL` must be negative, or am I understanding it incorrectly?

Comment (cjihrig) [reply src/uv-common.c]: Sorry, you're right. `UV__ERR()` only negates errors on platforms where they are positive.

File: docs/src/udp.rst [hunk 7]
```diff
@@ -150,6 +150,45 @@ API
 
     :returns: 0 on success, or an error code < 0 on failure.
 
+.. c:function:: int uv_udp_connect(uv_udp_t* handle, const struct sockaddr* addr)
+
+    It associates the UDP handle to a remote address and port, so every
```
Comment (cjihrig) [inline docs/src/udp.rst]: I think you can drop "It" at the beginning of the sentence.

File: docs/src/udp.rst [hunk 8]
```diff
@@ -150,6 +150,45 @@ API
...
+    message sent by this handle is automatically sent to that destination.
+    Calling this function with a `NULL` `addr`, disconnects the handle.
```
Comment (cjihrig) [inline docs/src/udp.rst]: No need for the comma after addr.

File: docs/src/udp.rst [hunk 9]
```diff
@@ -150,6 +150,45 @@ API
...
+    Trying to call `uv_udp_connect()` on an already connected handle will result
+    on an `UV_ISCONN` error.
```
Comment (cjihrig) [inline docs/src/udp.rst]: on -> in

File: docs/src/udp.rst [hunk 10]
```diff
@@ -150,6 +150,45 @@ API
...
+    Trying to disconnect a not connected handle will return an `UV_ENOTCONN`
```
Comment (cjihrig) [inline docs/src/udp.rst]: a not connected handle -> a handle that is not connected

File: src/uv-common.c [hunk 11]
```diff
@@ -284,24 +284,87 @@ int uv_tcp_connect(uv_connect_t* req,
...
-                uv_udp_send_cb send_cb) {
+int uv_udp_connect(uv_udp_t* handle, const struct sockaddr* addr) {
   unsigned int addrlen;
 
   if (handle->type != UV_UDP)
     return UV_EINVAL;
 
+  /* Disconnect the handle */
+  if (addr == NULL) {
+    if (!(handle->flags & UV__UDP_CONNECTED))
```
Comment (cjihrig) [inline src/uv-common.c]: Would it make sense to do this check in `uv__udp_disconnect()`?

File: src/uv-common.c [hunk 12]
```diff
@@ -284,24 +284,87 @@ int uv_tcp_connect(uv_connect_t* req,
...
+  }
+
   if (addr->sa_family == AF_INET)
     addrlen = sizeof(struct sockaddr_in);
   else if (addr->sa_family == AF_INET6)
     addrlen = sizeof(struct sockaddr_in6);
   else
     return UV_EINVAL;
 
+  if (handle->flags & UV__UDP_CONNECTED)
```
Comment (cjihrig) [inline src/uv-common.c]: Similarly, this could live in `uv__udp_connect()`.

File: test/test-udp-open.c [hunk 13]
```diff
@@ -215,3 +216,84 @@ TEST_IMPL(udp_open_twice) {
...
+  ASSERT(r == 0);
+
+  r = uv_udp_bind(&server, (const struct sockaddr*) &addr, 0);
+  ASSERT(r == 0);
+
+  r = uv_udp_recv_start(&server, alloc_cb, recv_cb);
+  ASSERT(r == 0);
+
+  r = uv_udp_send(&send_req,
+  	&client,
```
Comment (cjihrig) [inline test/test-udp-open.c]: Line up the function arguments?

File: test/test-udp-open.c [hunk 14]
```diff
@@ -215,3 +216,84 @@ TEST_IMPL(udp_open_twice) {
...
+  	&buf,
+  	1,
+  	NULL,
+  	send_cb);
+  fprintf(stderr, "r: %d\n", r);
```
Comment (cjihrig) [inline test/test-udp-open.c]: Is this intended?

File: docs/src/udp.rst [hunk 15]
```diff
@@ -150,6 +150,45 @@ API
 
     :returns: 0 on success, or an error code < 0 on failure.
 
+.. c:function:: int uv_udp_connect(uv_udp_t* handle, const struct sockaddr* addr)
+
+    Associate the UDP handle to a remote address and port, so every
+    message sent by this handle is automatically sent to that destination.
+    Calling this function with a `NULL` `addr` disconnects the handle.
+    Trying to call `uv_udp_connect()` on an already connected handle will result
+    in an `UV_ISCONN` error.
```
Comment (saghul) [inline docs/src/udp.rst]: UV_EISCONN?

File: docs/src/udp.rst [hunk 16]
```diff
@@ -150,6 +150,45 @@ API
...
+        with the address and port to associate to.
+
+    :returns: 0 on success, or an error code < 0 on failure.
+
+    .. versionadded:: 1.21.0
+
+.. c:function:: int uv_udp_getpeername(const uv_udp_t* handle, struct sockaddr* name, int* namelen)
+
+    Get the remote IP and port of the UDP handle on connected UDP handles.
+    On unconnected handles, it returns `UV_NOTCONN`.
```
Comment (saghul) [inline docs/src/udp.rst]: UV_ENOTCONN

File: src/unix/tcp.c [hunk 17]
```diff
@@ -279,44 +279,28 @@ int uv_tcp_open(uv_tcp_t* handle, uv_os_sock_t sock) {
 int uv_tcp_getsockname(const uv_tcp_t* handle,
                        struct sockaddr* name,
                        int* namelen) {
-  socklen_t socklen;
 
   if (handle->delayed_error)
     return handle->delayed_error;
 
-  if (uv__stream_fd(handle) < 0)
```
Comment (saghul) [inline src/unix/tcp.c]: is it OK to remove this?

File: src/uv-common.h [hunk 18]
```diff
@@ -61,13 +61,15 @@ extern int snprintf(char*, size_t, const char*, ...);
 
 #ifndef _WIN32
 enum {
+  UV__UDP_CONNECTED   = 0x200000,
```
Comment (saghul) [inline src/uv-common.h]: I guess this PR needs a rebase now that flags have been unified?

File: src/win/udp.c [hunk 19]
```diff
@@ -880,6 +906,50 @@ int uv__udp_bind(uv_udp_t* handle,
...
+      bind_addr = (const struct sockaddr*) &uv_addr_ip6_any_;
+    else
+      return UV_EINVAL;
+
+    err = uv_udp_maybe_bind(handle, bind_addr, addrlen, 0);
+    if (err)
+      return uv_translate_sys_error(err);
+  }
+
+  err = connect(handle->socket, addr, addrlen);
```
Comment (saghul) [inline src/win/udp.c]: is this always going to work, even in non-blocking mode?

Comment (santigimeno) [reply src/win/udp.c]: Yeah, this is not correct. Will look into it. Thanks

File: src/unix/tcp.c (hunk 17 above)
Comment (santigimeno) [reply src/unix/tcp.c]: I think so. It should be dealt by [this call](https://github.com/santigimeno/libuv/blob/f8fd063f6849a818863becaf0e63e02799dd0253/src/unix/core.c#L1358) tu `uv_fileno()`

File: src/win/udp.c (hunk 19 above)
Comment (santigimeno) [reply src/win/udp.c]: After a chat I had with @saghul while looking at the [MSDN documentation](https://docs.microsoft.com/en-us/windows/desktop/api/winsock2/nf-winsock2-connect) it was clear to us that `connect()` on Windows should always return immediately, so the code as it is should be fine. The specific quote is:
>For a connectionless socket (for example, type SOCK_DGRAM), the operation performed by connect is merely to establish a default destination address that can be used on subsequent send/ WSASend and recv/ WSARecv calls

**Comment by santigimeno** [comment]
PR updated to address latest comments

File: docs/src/udp.rst [hunk 20]
```diff
@@ -150,6 +150,45 @@ API
...
     :returns: 0 on success, or an error code < 0 on failure.
 
+.. c:function:: int uv_udp_connect(uv_udp_t* handle, const struct sockaddr* addr)
+
+    Associate the UDP handle to a remote address and port, so every
+    message sent by this handle is automatically sent to that destination.
+    Calling this function with a `NULL` `addr` disconnects the handle.
+    Trying to call `uv_udp_connect()` on an already connected handle will result
+    in an `UV_EISCONN` error.
+    Trying to disconnect a handle that is not connected will return an
```
Comment (cjihrig) [inline docs/src/udp.rst]: Should this start on the previous line?

File: docs/src/udp.rst [hunk 21]
```diff
@@ -150,6 +150,45 @@ API
...
+
+    :param handle: UDP handle. Should have been initialized with
+        :c:func:`uv_udp_init`.
+
+    :param addr: `struct sockaddr_in` or `struct sockaddr_in6`
+        with the address and port to associate to.
+
+    :returns: 0 on success, or an error code < 0 on failure.
+
+    .. versionadded:: 1.22.0
```
Comment (cjihrig) [inline docs/src/udp.rst]: All of the versions will need to be bumped.

File: src/unix/pipe.c [hunk 22]
```diff
@@ -234,10 +231,13 @@ static int uv__pipe_getsockpeername(const uv_pipe_t* handle,
 
   addrlen = sizeof(sa);
   memset(&sa, 0, addrlen);
-  err = func(uv__stream_fd(handle), (struct sockaddr*) &sa, &addrlen);
+  err = uv__getsockpeername((const uv_handle_t*) handle,
+                             func,
```
Comment (cjihrig) [inline src/unix/pipe.c]: Tiny nit: I think this and the following lines are indented one space too much.

File: src/unix/udp.c [hunk 23]
```diff
@@ -227,9 +227,14 @@ static void uv__udp_sendmsg(uv_udp_t* handle) {
...
     memset(&h, 0, sizeof h);
-    h.msg_name = &req->addr;
-    h.msg_namelen = (req->addr.ss_family == AF_INET6 ?
-      sizeof(struct sockaddr_in6) : sizeof(struct sockaddr_in));
+    if (req->addr.ss_family == AF_UNSPEC) {
+      h.msg_name = NULL;
+      h.msg_namelen = 0;
+    } else {
+      h.msg_name = &req->addr;
+      h.msg_namelen = (req->addr.ss_family == AF_INET6 ?
```
Comment (cjihrig) [inline src/unix/udp.c]: Unnecessary parens around everything?

**Comment by santigimeno** [comment]
Updated with latest comments.
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/963/
      https://ci.nodejs.org/view/libuv/job/libuv-in-node/43/

**Comment by santigimeno** [comment]
The failures in `libuv-in-node` are in this [test](https://github.com/nodejs/node/blob/master/test/parallel/test-dgram-address.js#L74_L81), because of [this change](https://github.com/libuv/libuv/pull/1872/files#diff-7364d102ee26a9359694e2fbdb5c4397L857) (now we return `UV_EBADF` instead of `UV_EINVAL`), which does not seem important. What do you think?

**Comment by cjihrig** [comment]
The Windows failures in the libuv CI look legitimate.

I don't think the `UV_EBADF` vs. `UV_EINVAL` is too important (and does address a TODO), but it's probably technically  semver-major. I won't hold this up on that detail. We should probably just bite the bullet and release v2 anyway though.

**Comment by santigimeno** [comment]
I hadn't looked into the Windows failures 🤦‍♂️ . It should be fixed now. Thanks for pointing that out.

Regarding to the error returned issue I'm fine either way. I we decide it's server major I'll target this PR to master.

WRT to `v2` I'm inclined to think we should do it but let's see if we can merge first the few pending PR's left and then decide.

**Comment by santigimeno** [comment]
New CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/964/
https://ci.nodejs.org/view/libuv/job/libuv-in-node/44/console

**Comment by 1st1** [comment]
> WRT to v2 I'm inclined to think we should do it but let's see if we can merge first the few pending PR's left and then decide.

Sorry for being "that guy", but what's the approximate ETA on this PR landing to libuv and getting released?  Will we have to wait for libuv v2 for it?  Context: UDP is currently completely broken in uvloop and I just need to make a decision of either forking libuv or waiting for it to get this PR in.

**Comment by cjihrig** [comment]
@santigimeno do you want to move forward with this?

**Comment by santigimeno** [comment]
It depends on whether we consider the `UV_EBADF` vs. `UV_EINVAL` change semver-major or just a bugfix :man_shrugging:. Personally I don't care if I have to target this to master and honestly it looks to me like the more correct thing to do, but if there's a good argument for the bugfix I think the PR it's ready to go.

**Comment by 1st1** [comment]
And what's the ETA for 2.x?

**Comment by cjihrig** [comment]
ASAP at this point. We've still got [a few PRs](https://github.com/libuv/libuv/pulls?q=is%3Aopen+is%3Apr+label%3Av2) to get through though.

**Comment by santigimeno** [comment]
PR re-targeted to master.
New CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1010/ looks good. The only failure seems like problem in the bot (it's currently failing with `v1.x` and `master` too).

**Comment by santigimeno** [comment]
New CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1032/

**Comment by santigimeno** [comment]
Failures are unrelated. Landing

**Comment by santigimeno** [comment]
Landed in 
https://github.com/libuv/libuv/commit/bf6c5be84d08c13dc248e7235ec17491706433ab, 
https://github.com/libuv/libuv/commit/050c8d53ce3cdea1f5f7342ee9da25ba7fd054e6 and 
https://github.com/libuv/libuv/commit/70d437130393dc8d9d01f2758b9b3c23a4dccedb. Thanks!

**Comment by cjihrig** [comment]
Nice job @santigimeno

**Comment by 1st1** [comment]
Thank you so much guys.  Now waiting for the v2!  🚀🚀🚀

**Comment by pspacek** [comment]
Hello! Are there any plans on releasing v2 any time soon? I would love to have connected UDP sockets available...

**Comment by bnoordhuis** [comment]
@pspacek See discussion in #1597. Short answer: no, not really. :)

**Comment by 1st1** [comment]
Can we reconsider and land this to the 1.x branch?  It seems that this PR is backwards compatible.  Waiting a couple of years for 2.x is, of course, doable, but having this new functionality *today* would result in a significantly faster UDP support in uvloop.

**Comment by cjihrig** [comment]
Is the `UV_EBADF` to `UV_EINVAL` change the only "breaking" change? If so, I'd personally be OK with it landing on 1.x.

**Comment by santigimeno** [comment]
> Is the `UV_EBADF` to `UV_EINVAL` change the only "breaking" change?

Yes, I think so. 
I'd be ok too but can't this be a problem on nodejs-land? The breaking change was uncovered by a failing nodejs [test](https://github.com/nodejs/node/blob/master/test/parallel/test-dgram-address.js#L74_L81), and if it's considered a breaking change there, this could be a problem as they would not be able to release minor/patch releases with future `v1.x` libuv versions.

/cc (active nodejs maintainers) @cjihrig @addaleax @bnoordhuis @bzoz

**Comment by saghul** [comment]
I'm -0 on this. They are both failures, but applications which may rely on the actual error could break. Not sure we want to open that can of worms.

**Comment by pspacek** [comment]
From my application-developer perspective I do not see how could sane code depend on EINVAL (in situations changed in this MR). If it was my code I would think I have a logic error somewhere.

**Comment by cjihrig** [comment]
> I'm -0 on this. They are both failures, but applications which may rely on the actual error could break. Not sure we want to open that can of worms.

@saghul I do agree with you. I'm worried about the `master` branch becoming a place that things like this go to die.

**Comment by saghul** [comment]
> I'm worried about the `master` branch becoming a place that things like this go to die.

Back when I had more time my thoughts were that `master` wouldn't graduate to `v2` unless we stopped spending time landing everything in `v1` and added Good Stuff (TM) onto `v2` alone. That would give people a good motivation for moving forward.

I do, however, also understand the other angle: people need / want this now.

Choices, choices!

Let's backport it and 🤞  so hell doesn't break loose. API breakeage is debatable since they are both errors...

**Comment by santigimeno** [comment]
PR with the backport: https://github.com/libuv/libuv/pull/2217

**Pull request closed without merging**